### PR TITLE
Release Links 0.9.3

### PR DESCRIPTION
**Pull request by dhil**
Minor release for Links (this time without links-mysql8 included). Various stability and improvement fixes.

### CODE REVIEW

**Comment by dhil** [comment]
Pardon my ignorance. I wonder whether I need to actively do something to get this package merged? Thanks in advance.

**Comment by avsm** [comment]
i think you need a constraint on the precise version of links for the various database packages. If you look at the CI logs, it fails on:

```

#=== ERROR while compiling links-mysql.0.9.2 ==================================#
# context              2.0.8 | linux/x86_64 | ocaml-base-compiler.4.12.0 | file:///src
# path                 ~/.opam/4.12/.opam-switch/build/links-mysql.0.9.2
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p links-mysql -j 47
# exit-code            1
# env-file             ~/.opam/log/links-mysql-1134-e7966f.env
# output-file          ~/.opam/log/links-mysql-1134-e7966f.out
### output ###
#       ocamlc database/mysql-driver/.links_mysql.objs/byte/links_mysql__Mysql_database.{cmi,cmo,cmt} (exit 2)
# (cd _build/default && /home/opam/.opam/4.12/bin/ocamlc.opt -w -40 -strict-formats -strict-sequence -safe-string -bin-annot -w A-4-42-44-45-48-60-67 -safe-string -dtypes -w Ae-44-45-60 -g -cclib -lunix -thread -g -bin-annot -I database/mysql-driver/.links_mysql.objs/byte -I /home/opam/.opam/4.12/lib/links/core -I /home/opam/.opam/4.12/lib/mysql -no-alias-deps -open Links_mysql -o database/mysql-driver/.links_mysql.objs/byte/links_mysql__Mysql_database.cmo -c -impl database/mysql-driver/mysql_database.ml)
# File "database/mysql-driver/mysql_database.ml", line 244, characters 10-24:
# 244 |   inherit Value.database
#                 ^^^^^^^^^^^^^^
# Error: This class expression is not a class structure; it has type
#        Links_core.Sql.printer -> Links_core.Value.database

```

You can constraint links-mysql via `"links" {=version}` to make it match the version of links to the current version of links-mysql. This will ensure opam updates all the users of links systematically and not give an older version of links and a newer links-mysql.

**Comment by dhil** [comment]
Thanks @avsm!

**Comment by avsm** [comment]
Bonus points if you also push the same constraint fixes to older links-* packages :-) links-mysql.0.9.2 was selecting links.0.9.3 in the failing CI, for example

**Comment by dhil** [comment]
Thanks again @avsm. I think I constrained all the packages now. I have opened a PR links-lang/links#973 to constrain future releases --- I'll merge it soon.

File: packages/links-mysql/links-mysql.0.9.3/opam [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+authors: "The Links Team <links-dev@inf.ed.ac.uk>"
+synopsis: "MySQL database driver for the Links Programming Language"
+description: "MySQL database driver for the Links Programming Language"
+homepage: "https://github.com/links-lang/links"
+dev-repo: "git+https://github.com/links-lang/links.git"
+bug-reports: "https://github.com/links-lang/links/issues"
+license: "GPL-2"
+
+build: [
+  [ "dune" "subst" ] {pinned}
```
Comment (mseri) [inline packages/links-mysql/links-mysql.0.9.3/opam:12]:
```suggestion
  [ "dune" "subst" ] {dev}
```

File: packages/links-postgresql/links-postgresql.0.9.3/opam [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+synopsis: "Postgresql database driver for the Links Programming Language"
+description: "Postgresql database driver for the Links Programming Language"
+homepage: "https://github.com/links-lang/links"
+dev-repo: "git+https://github.com/links-lang/links.git"
+bug-reports: "https://github.com/links-lang/links/issues"
+license: "GPL-2"
+
+
+build: [
+  [ "dune" "subst" ] {pinned}
```
Comment (mseri) [inline packages/links-postgresql/links-postgresql.0.9.3/opam:13]:
```suggestion
  [ "dune" "subst" ] {dev}
```

File: packages/links-sqlite3/links-sqlite3.0.9.3/opam [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+authors: "The Links Team <links-dev@inf.ed.ac.uk>"
+synopsis: "SQLite database driver for the Links Programming Language"
+description: "SQLite database driver for the Links Programming Language"
+homepage: "https://github.com/links-lang/links"
+dev-repo: "git+https://github.com/links-lang/links.git"
+bug-reports: "https://github.com/links-lang/links/issues"
+license: "GPL-2"
+
+build: [
+  [ "dune" "subst" ] {pinned}
```
Comment (mseri) [inline packages/links-sqlite3/links-sqlite3.0.9.3/opam:12]:
```suggestion
  [ "dune" "subst" ] {dev}
```

File: packages/links/links.0.9.3/opam [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+maintainer: "Simon Fowler <simon.fowler@ed.ac.uk>"
+authors: "The Links Team <links-dev@inf.ed.ac.uk>"
+synopsis: "The Links Programming Language"
+description: "Links is a functional programming language designed to make web programming easier."
+homepage: "https://github.com/links-lang/links"
+dev-repo: "git+https://github.com/links-lang/links.git"
+bug-reports: "https://github.com/links-lang/links/issues"
+license: "GPL-2"
+build: [
+  [ "dune" "subst" ] {pinned}
```
Comment (mseri) [inline packages/links/links.0.9.3/opam:11]:
```suggestion
  [ "dune" "subst" ] {dev}
```

**Comment by mseri** [comment]
@dhil one last question and then I merge, is the license "GPL-2.0-only" or "GPL-2.0-or-later", do you know?

**Comment by dhil** [comment]
@mseri I asked this question a few days ago links-lang/links#974. I will push for answer and update the packages accordingly!

**Comment by mseri** [comment]
Thanks, this change can be done later since affects also older versions and the current license is not incorrect

**Comment by dhil** [comment]
Thanks! Once I have an answer, I will open a PR to retrospectively update the license field in each package.

**Pull request merged**